docs: started work on a type reference (#3590)
idris-lang/Idris2: docs/source/reference/types.rst

Types, Functions, and Interfaces
================================

The type system in Idris 2 is quite powerful, and is the main feature that sets it apart from simply typed languages like Haskell or OCaml.

There are many ways to construct types, those being members of ``Type``, in Idris.

-------------
Functions
-------------

> Function types, formed by ``(x : A) -> B``, are valid types, given that ``A`` is a valid type *or* kind and ``B`` is a valid type *or* kind. 

Reviewer: ```suggestion
Function types, formed by ``(x : a) -> b``, are valid types, given that ``a``  and ``b`` are valid types.
```

protip: use lower case for generic names for types since they are automatically quantified in the compiler. It should probably be written in this file that `map : (a -> b) -> List a -> List b` is actually `map : {0 a, b : Type} -> (a -> b) -> List a -> List b` because `a` and `b` are lowercase but `List` is uppercase so it's not automatically added as an argument.

Also Idris does not have kinds, only types.

As shorthand, for non-dependent types, ``A -> B`` is shorthand for ``(_ : A) -> B``.

^^^^^^^^^^^^^
Implicits
^^^^^^^^^^^^^

Implicit function types, those where the value of a parameter can be inferred from the context, can take three forms.
The first of these, ``{x : A} -> B``, forms the basis "basic" implicit types. 
That is, ``x`` should be inferred *syntactically*. 
This is the equivalent of implicit ``forall`` in Haskell or Rocq, where we might state that ``vectorMap`` on a vector might be

.. code:: idris

    vectorMap : {n : Nat} -> (a -> b) -> (Vect n a) -> (Vect n b)

We would then use ``vectorMap`` as follows: 

.. code-block:: idris

    vectorMap f v

or

.. code:: idris

    vectorMap {n = m} f v

These two examples show the power of Idris implicits: while they can be inferred by the compiler, if necessary (or desired) they may be specified by the user

.. tip:: 

>     Idris *not have* a ``forall`` keyword, unlike Haskell. 

Reviewer 2: Idris does have a `forall`, but I don't know if it differs in meaning from that in Haskell

Reviewer 2: `forall x . <stuff>` is `{0 x : _} -> <stuff>`

.. warning::
    If you want to use a type name to specify a value in Idris, you *cannot* merely use it as an implicit.
    This is because, as stated before, instead of ``forall``'s we have implicits. 
    This means that if you want to use a type variable, say ``a`` in ``castOne : Cast a b => a -> [b]`` in the function, you must explicitly take it as an argument, e.g ``castOne : {a : Type} -> Cast a b => a -> [b]``, and then use it as ``castOne {a} x``

^^^^^^^^^^^^^
Quantities
^^^^^^^^^^^^^

Function types may also give quantities to their arguments.
These are either ``0``, ``1``, or nothing (many), which are all placed before a named parameter, so, we have ``(# x : A)``.
A value with a multiplicity of zero (erased) must only be used in the type signature (i.e, a Haskell type variable).
A value with a multiplicity of one (linear) must be used exactly once in the definition of the value.

---------------------------------------
Interfaces and Implementations
---------------------------------------

Interfaces are the Idris equivalent of Haskell type-classes.
They are modeled as a lot of sugar on implicit types.

> There syntax is as follows:

Reviewer 2: ```suggestion
Their syntax is as follows:
```

.. code:: idris

    vectorMap : {n : Nat} -> (a -> b) -> Vect n a -> Vect n b

We would then use ``vectorMap`` as follows:

.. code-block:: idris

    vectorMap f v

or

.. code:: idris

    vectorMap {n = m} f v

These two examples show the power of Idris implicits: while they can be inferred by the compiler, if necessary (or desired) they may be specified by the user

Idris has a *two* additional modifiers of implicits, however.
The first, while not that important to the type system of Idris, is *incredibly* useful, and it is the default implicit.
Essentially "regular" implicits only try to figure out the type without adding any extra information.

Default implicits, on the other hand, instead of trying to figure out a value, will instead just use a default, which means that unnecessary "simple" functions are unnecessary in Idris: you just use the default value.

The syntax for default implicits is ``{default val x : a} -> b``, where ``val : a``, which is the default value for ``x``.

The third kind of implicits are auto implicits.
Auto implicits are most useful in proofs, where they differ from regular implicits by irrelevance.

> That is, a regular default will never initiate *anything* that cannot otherwise be initiated, while an ``auto`` default will not change anything outside of the input that cannot be initiated.

Reviewer 2: what does the second half of this sentence mean?

In QTT, every function type has an associated multiplicity, ``0``, ``1``, or ω.
These correspond with how many times the given input is "used" in the result type.

Very roughly, ``0`` corresponds to a value "being ignored" or *erased*, a value of ``1`` corresponds to a value being "used" once, or being *one-to-one*, and ω (which, in Idris syntax is ommited) corresponds to a type having an unrestricted number of uses.

Note that this idea of "usage" is not always cut and dry.
For instance, as we saw above, ``{0 a : Type} -> {0 b : Type} -> (a -> b) -> List a -> List b`` is a valid type, despite the fact it uses an erased arguement.
This is because of the fact that erased arguements are only ignored at runtime, however, in the above, the usages of ``a`` and ``b`` are in a type, so they are fine.
In addition, we can alias values without changing their multiplicity, so long as we treat them as the same value.
For instance, if ``1 x : a``, ``let 1 y = x in EXPR`` is valid so long as exactly *one of* ``x`` or ``y`` is used in ``EXPR``.

The syntax for specifying multiplicity in functions is ``(# x : a) -> B`` (or any variation on such), where ``#`` is the multiplicty.

> In addition, we may also declare ``let`` binding with a multiplictity, with the syntax ``let # x : a = val in EXPR`` or ``let # x = val in EXPR``

Reviewer 2: I didn't quite mean to add the mention of `let`, since there are a number of other places where you'd do this (e.g. in a "type alias")
```idris
0 Numbers : Type
Numbers = List Int
```
I more mentioned it to point out that function parameters isn't the only place this happens

.. note::
    "Using" a value of multiplicity 1 changes the multiplicity to 0

---------------------------------------
Interfaces and Implementations
---------------------------------------

.. error::
    Add info on interface constructors

Interfaces are the Idris equivalent of Haskell type-classes.
They are modeled as a lot of sugar on implicit types.

> There syntax is as follows:

Reviewer 2: ```suggestion
Their syntax is as follows:
```

In addition, to do simple universal quantification, one can write ``forall x . a``, where ``x`` must be only a name (it can't have a type qualifier).
This is desugared to ``{0 x : _} -> a``.

^^^^^^^^^^^^^
Implicits
^^^^^^^^^^^^^

Implicit function types, those where the value of a parameter can be inferred from the context, can take three forms.
The first of these, ``{x : A} -> B``, forms the basis "basic" implicit types.
That is, ``x`` should be inferred *syntactically*.
This is the equivalent of implicit ``forall`` in Haskell or Rocq, where we might state that ``vectorMap`` on a vector might be:

.. code:: idris

    vectorMap : {n : Nat} -> (a -> b) -> Vect n a -> Vect n b

We would then use ``vectorMap`` as follows:

.. code-block:: idris

    vectorMap f v

or

.. code:: idris

    vectorMap {n = m} f v

These two examples show the power of Idris implicits: while they can be inferred by the compiler, if necessary (or desired) they may be specified by the user

Idris has *two* additional modifiers of implicits, however.
The first, while not that important to the type system of Idris, is *incredibly* useful, and it is the default implicit.

> Essentially "regular" implicits only try to figure out the type without adding any extra information.
> 
> Default implicits, on the other hand, instead of trying to figure out a value, will instead just use a default, which means that unnecessary "simple" functions are unnecessary in Idris: you just use the default value.

Reviewer: ```suggestion
Where "regular" implicits only try to resolve the type without adding any extra information, default implicit arguments, will instead use a default value when the value cannot be resolved syntactically.

This is *incredibly* useful to avoid repeating the same function definition multiple times with a different set of arguments when the most common case can be given via a default implicit and the other uses can give the argument by name.
```

I tried to combine you previous sentence about how useful it is with the one explaning how to use default implicits.

The syntax for default implicits is ``{default val x : a} -> b``, where ``val : a``, which is the default value for ``x``.

The third kind of implicits are auto implicits.

> Auto implicits are most useful in proofs, where they differ from regular implicits by there purpose.

Reviewer: That's somewhat debatable. I suggest removing this sentence and instead show multiple uses for auto-implicits:

- proofs (like `Elem` or `IsJust`)
- interfaces
- effects (like how the compiler has tags to track allowable effects)

> Regular implicits try to find a "necessary" match.
> ``auto`` implicits, on the other hand, look for *any* possible construction of a value.

Reviewer: ```suggestion
Regular implicits try to find a "necessary" match that is in scope.
 ``auto`` implicits, on the other hand, look for *any* possible construction of a value, even if it is not in scope, the compiler will attempt to construct a value for it.
```

> This is useful for propositions, where we often don't need to reason about the specific value of a type member, merely that it exists.

Reviewer: ```suggestion
This is useful for properties of types where we don't want to change the definition of the type but carry along some facts about it.
```

Not quite sure what the original sentence meant but I thought this was a more illuminating use-case for auto implicits. It's often the case that it's the best way to carry along a proof from a library type, because building your own is a hassle.

> Essentially, auto implicits are for cases where we want the compiler to have more freedom to infer the value

Reviewer: ```suggestion
```
This sentence needs to go. This is actually a poor use for auto implicits since they are unpredictable and the compiler will generate different values for the same type depending on constructor definition order and other compiler internal details. It is much better to provide it with a decision procedure with a unary constructor than use it as a way to get more leeway out of the term-inference machinery.

For instance, if we want to automatically pass a proof through some functions, we use ``auto`` defaults.

.. hint::
    Lowercase names in types are automatically universally quantified in types.
    For instance, ``map : (a -> b) -> List a -> List b`` is transformed into ``map : {0 a : Type} -> {0 b : Type} -> (a -> b) -> List a -> List b`` [#f1]_.
    You can write this signature explicitly, and it will compile the same as the simple version, and in addition you could also have the ``forall`` keyword and wrote this as ``map : forall a . forall b . ((a -> b) -> List a -> List b)``.

    If desired, this behavior can be toggled off by ``%unbound_implicits off`` (and back on by ``%unbound_implicits on``)

^^^^^^^^^^^^^
Quantities
^^^^^^^^^^^^^

Idris 2 uses Quantitative Type Theory (`QTT`_) as its theoretical basis.

> In QTT, every function type has an associated multiplicity, ``0``, ``1``, or ω.

Reviewer: ```suggestion
In QTT, every variable has an associated multiplicity, ``0``, ``1``, or ω.
```

Essentially every binder  `a : b` has a quantity attached to them: lambdas, pi-types, record fields, toplevel definitions, etc. probably important to explain that in some way. "Variable" is probably good enough for an entry-level document

These correspond with how many times the given input is "used" in the result type.

> Very roughly, ``0`` corresponds to a value "being ignored" or *erased*, a value of ``1`` corresponds to a value being "used" once, or being *one-to-one*, and ω (which, in Idris syntax is omitted) corresponds to a type having an unrestricted number of uses.

Reviewer: ```suggestion
Very roughly, ``0`` corresponds to a value existing only during typechecking but not at runtime. 
A value of ``1`` corresponds to a value being used exactly once, and ω (which, in Idris syntax is omitted) corresponds to a type having an unrestricted number of uses.
```

> Note that this idea of "usage" is not always cut and dry.

Reviewer: Usage is cut and dry. If your variable appears at runtime then it's used, erased quantities cannot appear at runtime, period.

The only loose rules are around subtyping of ω arguments and linearity but that shouldn't be something the audience of this text needs to worry about.

For instance, as we saw above, ``{0 a : Type} -> {0 b : Type} -> (a -> b) -> List a -> List b`` is a valid type, despite the fact it uses an erased argument.
This is because erased arguments are only ignored at runtime, however, in the above, the usages of ``a`` and ``b`` are in a type, so they are fine.
In addition, we can alias values without changing their multiplicity, so long as we treat them as the same value.

> For instance, if ``1 x : a``, ``let 1 y = x in EXPR`` is valid so long as exactly *one of* ``x`` or ``y`` is used in ``EXPR``.

Reviewer: that is not true. Only `y` can be use in the body `EXPR`. `x` has been "spent" by being bound to `y`. And you can see this by writing a small program:

```
fn : (1 x : Nat) -> IO ()
fn x = let 1 y = x in ?hole
```

The hole has context:
```
 0 x : Nat
 1 y : Nat
------------------------------
hole : IO ()
```

This is one of many examples of how the quantitative notion of "usage" is not always exactly the same as the intuitive notion of "usage".

The syntax for specifying multiplicity in functions is ``(# x : a) -> B`` (or any variation on such), where ``#`` is the multiplicity.

> In addition, we may also declare ``let`` binding with a multiplicity, with the syntax ``let # x : a = val in EXPR`` or ``let # x = val in EXPR``.

Reviewer: ```suggestion
In addition, we may also declare ``let`` binding with a multiplicity, with the syntax ``let qty x : a = val in expr`` or ``let qty x = val in expr``.
```

``data`` declares (General) Algebraic Data Types ((G)ADTs).
The Haskell style, simple ADT form is always a statement of the form ``data NAME ARGS = ALT {| ALT}``.
``ARGS`` must be a valid list of names.
In addition, each ``ALT`` must be of the form ``CONS PARAMS`` where each ``CONS`` is either an operator or a name (that will be declared as a constructor).
Finally, each of ``PARAMS`` must be a valid type.

To declare (Agda style) Generalized Algebraic Data Types, we must write ``data NAME : KIND where CONS``.
``KIND`` must be a kind (in Idris there isn't actually distinction between types and kinds) whose tail ultimately ends up returning the type ``Type``.
So, for instance, ``data Vect : (n : Nat) -> Type where ...`` is valid, whilst ``data Vect : forall k. (n : Nat) -> k where ...`` is not.
Each ``CONS`` must be a function declaration, except that the name being defined must start with an uppercase letter and the ultimate return type of it must be a fully applied instance of ``NAME``.

The other way to define top level types, ``record``, is documented in detail on the `records`_ page.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Existentials in Types

Reviewer: ```suggestion
Inferable Placeholders
```

I don't know what to call this class of terms but `existential` is extremely confusing because there is already something called `existential types` in programming and it's not what is exposed here. Existential types are usually an encoding of sigma-types for languages without dependent types.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

While not restricted to types, existentials are so prevalent within types that there usage bears mentioning.
Existentials are always of the form ``_``, ``?`` or ``?NAME``, where ``NAME`` is a valid name.

> Named existentials are scoped by *module*, however, they may only exist in one term.

Reviewer 3: I don't think ? is like an auto-implicit (you need %search for that). I'm pretty sure it's solved by inference. The difference between _ and ? is that ? can be solved later. If you use _ in a type signature it's an error if it is not solved at the end of a signature. If you use ?, it can be solved later in the file. (An error if not solved by the end of the file.)

.. caution::

>     Existentials are scoped over modules.

Reviewer: ```suggestion
    Holes are scoped over modules.
```

    This means that a given hole must have a unique name in a module.

Named existential forms an equivalent to "typed holes", that is, a value to be filled in later.
These can occur in types, although typically occur in generated ones.

> Most often, this will occur in usages of generic types, where, for instance, given ``show : {a : Type} -> Show a => a -> String``, we transform the context ``show : {a : Type} -> Show a => a -> String, y = show x |- y : String`` to ``x : ?a, Show ?a |- y : String`` [#f2]_ .

Reviewer: I think this needs to be reworked in light of the previous changes

> ``?`` is similar except, while if ``?a``` cannot be inferred it is made it a typed hole, if ``?`` cannot be inferred, it raises an error

Reviewer: ```suggestion
```

`?` is completely different as it infers a value in addition to the type. Holes only infer the type.

1. Take the interface, turn it into a record, and make all its methods into arguments to the record
2. Turn every instance of the interface into a definition of the record.
3. For each named one, assign the name to the definition, for each unnamed one, add the ``%hint`` pragma
4. Transform every case of ``f : i a => b`` to ``f : {_ : i' a} -> b``, and pattern match on the implicit argument in the definition.

Given this, it seems natural to ask how we can construct "anonymous" inline instances.
Idris takes this into account, by providing interface constructors.
An interface constructor extends an interface by allowing a ``constructor NAME`` clause in an interface definition, which creates a "interface value" constructor. For instance, the following two implementations are equivalent (given `semigroup`_):

.. code:: idris

    [semiList] {a : Type} -> Semigroup (List a) where
        (<+>) = (++)

and

.. code:: idris

    semiList : {a : Type} -> Semigroup (List a)
    semiList = (MkSemigroup (++))

----------------------
Other Forms
----------------------

The two main ways to declare types at the top level are ``record`` and ``data``

.. note::

>     Type synonyms, type families, and the like from Haskell are all just normal functions from ``Type`` to something else.

Reviewer 4: Actually, aren't these functions that _result_ in a type. For instance, a type synonym is just

```idris
Label : Type
Label = String
```

Same for type families? In fact, we hardly ever write functions of the type `Type -> a`.

>     Type synonyms, type families, and the like from Haskell are all just normal functions from ``Type`` to something else.

Reviewer: Type families both result in a type and also depend on types given as argument. open type families is closest to pattern matching on types, data families are like an indexed interface `a -> Type`. None of those comparaisons actually make any sense if you think too hard about it. Maybe the paragraph should be scraped because there isn't much pedagogical benefit in relating dependent types to haskell-specific features that don't quite match

    In addition, this means that one can have anonymous instances, so, for instance, ``Monad (\x => x)`` is a perfectly valid form.

    Even so, it is in general not a good idea to create such instances

``data`` declares (General) Algebraic Data Types ((G)ADTs).
The Haskell style, simple ADT form is always a statement of the form ``data NAME ARGS = ALT {| ALT}``.
``ARGS`` must be a valid list of names.
In addition, each ``ALT`` must be of the form ``CONS PARAMS`` where each ``CONS`` is either an operator or a name (that will be declared as a constructor).
Finally, each of ``PARAMS`` must be a valid type.

> To declare (Agda style) Generalized Algebraic Data Types, we must write ``data NAME : KIND where CONS``.

Reviewer: I found the post, quite amusing for us. Not amusing for them
